x/evm/keeper/msg_server.go

```diff
@@ -55,6 +55,285 @@ func validateChainActivated(ctx sdk.Context, n types.Nexus, chain nexus.Chain) e
+	}
+
+	voter := s.snapshotter.GetOperator(ctx, req.Sender)
+	if voter == nil {
+		return nil, fmt.Errorf("account %v is not registered as a validator proxy", req.Sender.String())
+	}
+
+	poll := s.voter.GetPoll(ctx, req.PollKey)
+	switch {
+	case poll.Is(vote.NonExistent):
+		return nil, fmt.Errorf("vote for non-existent poll %s", req.PollKey)
+	case poll.Is(vote.Expired):
+		return &types.VoteConfirmGatewayTxResponse{Log: fmt.Sprintf("vote for poll %s already expired", poll.GetKey())}, nil
+	case poll.Is(vote.Failed), poll.Is(vote.Completed):
+		return &types.VoteConfirmGatewayTxResponse{Log: fmt.Sprintf("vote for poll %s already decided", poll.GetKey())}, nil
+	}
+
+	if err := poll.Vote(voter, &req.Vote); err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(types.EventTypeGatewayTxConfirmation,
+			sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+			sdk.NewAttribute(sdk.AttributeKeyAction, types.AttributeValueVote),
+			sdk.NewAttribute(types.AttributeKeyPoll, string(types.ModuleCdc.MustMarshalJSON(&req.PollKey))),
+			sdk.NewAttribute(types.AttributeKeyValue, req.Vote.String()),
+		),
+	)
+
+	if poll.Is(vote.Pending) {
+		return &types.VoteConfirmGatewayTxResponse{Log: fmt.Sprintf("not enough votes to confirm poll %s yet", poll.GetKey())}, nil
+	}
+
+	if poll.Is(vote.Failed) {
+		return &types.VoteConfirmGatewayTxResponse{Log: fmt.Sprintf("poll %s failed", poll.GetKey())}, nil
+	}
+
+	result, ok := poll.GetResult().(*types.VoteConfirmGatewayTxRequest_Vote)
+	if !ok {
+		return nil, fmt.Errorf("result of poll %s has wrong type, expected VoteConfirmGatewayTxRequest_Vote, got %T", req.PollKey.String(), poll.GetResult())
+	}
+
+	s.Logger(ctx).Info(fmt.Sprintf("gateway transaction confirmation confirmation result is %s", result.String()), "chain", chain.Name, "pollKey", req.PollKey.String())
+
+	event := sdk.NewEvent(types.EventTypeGatewayTxConfirmation,
+		sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
+		sdk.NewAttribute(types.AttributeKeyChain, chain.Name),
+		sdk.NewAttribute(types.AttributeKeyTxID, req.GetTxID()),
+		sdk.NewAttribute(types.AttributeKeyPoll, string(types.ModuleCdc.MustMarshalJSON(&req.PollKey))))
+
+	if len(result.Events) == 0 {
+		poll.AllowOverride()
+		ctx.EventManager().EmitEvent(
+			event.AppendAttributes(sdk.NewAttribute(sdk.AttributeKeyAction, types.AttributeValueReject)))
+
+		return &types.VoteConfirmGatewayTxResponse{Log: fmt.Sprintf("poll %s failed to decide anything", poll.GetKey())}, nil
+	}
+
+	ctx.EventManager().EmitEvent(
+		event.AppendAttributes(sdk.NewAttribute(sdk.AttributeKeyAction, types.AttributeValueConfirm)))
+
+	queue := keeper.GetContractCallQueue(ctx)
+	txID := req.GetTxID()
+
+	for _, event := range result.Events {
+		key := utils.LowerCaseKey(txID).AppendStr(strconv.FormatUint(event.Info.Index, 10))
+		if queue.Has(key) {
+			return nil, fmt.Errorf("event %s-%d already confirmed", txID, event.Info.Index)
+		}
```

Comment:
Using the queue to check seems brittle. Once it's processed the key won't be found anymore. I think we should find a better way to check if an event is confirmed